cli/bench/lsp_bench_standalone.rs

```diff
@@ -0,0 +1,75 @@
+use deno_bench_util::bencher::benchmark_group;
+use deno_bench_util::bencher::benchmark_main;
+use deno_bench_util::bencher::Bencher;
+use deno_core::serde_json;
+use deno_core::serde_json::json;
+use deno_core::serde_json::Value;
+
+// Intended to match the benchmark in quick-lint-js
+// https://github.com/quick-lint/quick-lint-js/blob/35207e6616267c6c81be63f47ce97ec2452d60df/benchmark/benchmark-lsp/lsp-benchmarks.cpp#L223-L268
+fn incremental_change_wait(bench: &mut Bencher) {
+  let deno_exe = test_util::deno_exe_path();
+  let mut client = test_util::lsp::LspClient::new(&deno_exe).unwrap();
+
+  static FIXTURE_INIT_JSON: &[u8] =
+    include_bytes!("testdata/initialize_params.json");
+  let params: Value = serde_json::from_slice(FIXTURE_INIT_JSON).unwrap();
+  let (_, maybe_err) = client
+    .write_request::<_, _, Value>("initialize", params)
+    .unwrap();
+  assert!(maybe_err.is_none());
+  client.write_notification("initialized", json!({})).unwrap();
+
+  client
+    .write_notification(
+      "textDocument/didOpen",
+      json!({
+        "textDocument": {
+          "uri": "file:///testdata/express-router.js",
+          "languageId": "javascript",
+          "version": 0,
+          "text": include_str!("testdata/express-router.js")
+        }
+      }),
+    )
+    .unwrap();
+  let (method, maybe_diag): (String, Option<Value>) =
+    client.read_notification().unwrap();
+  assert_eq!(method, "textDocument/publishDiagnostics");
+  let _expected_num_diagnostics = get_num_diagnostics(maybe_diag);
+
+  let mut version = 0;
+  bench.iter(|| {
+      let text = format!("m{:05}", version);
+      client
+        .write_notification(
+          "textDocument/didChange",
+          json!({
+              "textDocument": {
+                  "version": version,
+                  "uri":"file:///testdata/express-router.js"
+              },
+              "contentChanges":[
+                {"text": text, "range":{"start":{"line":506,"character":39},"end":{"line":506,"character":45}}},
+                {"text": text, "range":{"start":{"line":507,"character":8},"end":{"line":507,"character":14}}},
+                {"text": text, "range":{"start":{"line":509,"character":10},"end":{"line":509,"character":16}}}
+              ]
+          })
+      ).unwrap();
+      let (method, maybe_diag): (String, Option<Value>) =
+        client.read_notification().unwrap();
```

Comment:
We have up to 3 publishes that occur when a change is made:

https://github.com/denoland/deno/blob/eda6e58520276786bd87e411d0284eb56d9686a6/cli/lsp/diagnostics.rs#L605-L612

When each of the three categories finish being calculated, they publish.  This code is only measuring from change to first publish, not full publish.  I am not sure what you are wanting here.

I realised when looking at the code again, we are debouncing "all" the diagnostics, when in fact we only care about debouncing the "typescript" ones because of their blocking nature for other API calls, which degraded performance:

https://github.com/denoland/deno/blob/eda6e58520276786bd87e411d0284eb56d9686a6/cli/lsp/diagnostics.rs#L172-L184

We could/should break that out, so that the "lint" and "deno (dependency)" diagnostics are calculated independently and not debounced.  There might be a little bit work to do, because we treat the dirtyness of the diagnostics per document, and what we are saying is that lint and deno diagnostics are one version of dirty and typescript is another version of dirty.

Reply:
Thanks, now I understand why quick-lint has `diagnosticsMessagesToIgnore` set to 2 for Deno.

https://github.com/quick-lint/quick-lint-js/blob/35207e6616267c6c81be63f47ce97ec2452d60df/benchmark/benchmark-lsp/benchmark-config.json#L63-L77